### PR TITLE
Add ContextStore

### PR DESCRIPTION
**Pull request by shaoshuai0102**

### CODE REVIEW

**Comment by shaoshuai0102** [comment]
ping @dead-horse

File: index.js [hunk 1]
```diff
@@ -78,10 +78,8 @@ function formatOpts(opts) {
     opts.decode = util.decode;
   }
 
-  if (opts.store) {
-    assert(typeof opts.store.get === 'function', 'store.get must be function');
-    assert(typeof opts.store.set === 'function', 'store.set must be function');
-    assert(typeof opts.store.destroy === 'function', 'store.destroy must be function');
+  if (opts.ContextStore) {
+    assert(typeof opts.ContextStore === 'function', 'ContextStore must be a function');
```
Comment (popomore) [inline index.js]: `is.class`

File: index.js [hunk 2]
```diff
@@ -78,10 +78,8 @@ function formatOpts(opts) {
     opts.decode = util.decode;
   }
 
-  if (opts.store) {
```
Comment (popomore) [inline index.js]: Should check store first when it's specified.

File: index.js (hunk 1 above)
Comment (popomore) [inline index.js]: And check `ContextStore.prototype[method]`

File: lib/context.js [hunk 3]
```diff
@@ -16,7 +17,13 @@ class ContextSession {
   constructor(ctx, opts) {
     this.ctx = ctx;
     this.opts = Object.assign({}, opts);
-    this.store = this.opts.store;
+    this.store = this.opts.ContextStore ? new this.opts.ContextStore(ctx) : this.opts.store;
+    if (this.store) {
+      assert(typeof this.store.get === 'function', 'store.get must be function');
+      assert(typeof this.store.set === 'function', 'store.set must be function');
+      assert(typeof this.store.destroy === 'function', 'store.destroy must be function');
```
Comment (fengmk2) [inline lib/context.js]: No need to check store type on every request.
Only assert once at the beginning.

File: index.js (hunk 1 above)
Comment (fengmk2) [inline index.js]: +1

Comment (fengmk2) [inline index.js]: only check `ContextStore.prototype[method]` once

File: test/store.test.js [hunk 4]
```diff
@@ -6,635 +6,654 @@ const should = require('should');
 const mm = require('mm');
 const session = require('..');
 const store = require('./store');
+const ContextStore = require('./context_store');
```
Comment (fengmk2) [inline test/store.test.js]: keep the old test file and add a new one for context store.

**Comment by shaoshuai0102** [comment]
All issues settled

**Comment by dead-horse** [comment]
should introduce `ContextStore` in readme

File: Readme.md [hunk 5]
```diff
@@ -118,6 +118,8 @@ app.use(convert(session(app)));
 
   Once you passed `options.store`, session is strong dependent on your external store, you can't access session if your external store is down. **Use external session stores only if necessary, avoid use session as a cache, keep session lean and stored by cookie!**
 
+  If your session store requires data or utilities from context, `opts.ContextStore` is alse supported. `ContextStore` must be a class which claims three instance methods demonstrated above. `new ContextStore(ctx)` will be executed on every request.
```
Comment (shaoshuai0102) [inline Readme.md]: Added

**Comment by dead-horse** [comment]
LGTM

**Comment by dead-horse** [comment]
@shaoshuai0102 please pick this feature to master branch

**Comment by shaoshuai0102** [comment]
publish a version pls.

I will pick it to master

**Comment by shaoshuai0102** [comment]
`4.2.0`

**Pull request merged**